### PR TITLE
Add support for image cutout by id for deepCoadd and raw images

### PR DESCRIPTION
**Pull request by kennywlo**
Please review the changes to support image cutouts by id for deepCoadd and raw images.

### CODE REVIEW

File: python/lsst/dax/imgserv/locateImage.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def getImageByDataId(self, ra, dec, width, height, qResults, cutoutType="arcseco
         xyWcs = wcs.skyToPixel(raDec)
         x0 = img.getX0()
         y0 = img.getY0()
-        xyCenter = afwGeom.Point2D(xyWcs.getX() - x0, xyWcs.getY() - y0)
+        xyCenter = afwGeom.Point2I(int(xyWcs.getX()-x0), int(xyWcs.getY()-y0))
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Needs spaces around '-'.

File: python/lsst/dax/imgserv/locateImage.py [hunk 2]
```diff
@@ -269,13 +269,16 @@ def _getImageButler(self, qResults):
         '''
         # This will return on the first result.
         log.debug("Raw_getImageButler qResults:{}".format(qResults))
-        for ln in qResults:
-            run, camcol, field, filterName = ln[2:6]
+        valid, run, camcol, field, filterName = _getKeysForButler(qResults)
+        if valid is True:
+            log.debug("Raw_getImageButler run={} camcol={} field={} filter={}".format(run,
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Line too long.

File: python/lsst/dax/imgserv/locateImage.py [hunk 3]
```diff
@@ -269,13 +269,16 @@ def _getImageButler(self, qResults):
...
-        for ln in qResults:
-            run, camcol, field, filterName = ln[2:6]
+        valid, run, camcol, field, filterName = _getKeysForButler(qResults)
+        if valid is True:
+            log.debug("Raw_getImageButler run={} camcol={} field={} filter={}".format(run, 
+                camcol, field, filterName))
             butler = lsst.daf.persistence.Butler(self._dataRoot)
-            img = butler.get(self._imageDatasetType, run=run, camcol=camcol,
-                    field=field, filter=filterName)
+            img = butler.get(self._imageDatasetType, run=run, camcol=camcol, field=field, filter=filterName)
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Line too long.

File: python/lsst/dax/imgserv/locateImage.py [hunk 4]
```diff
@@ -286,6 +289,22 @@ def _getMetadata(self, butler, qResults):
                               field=field, filter=filterName)
 
 
+    def _getImageCutoutFromScienceId(self, scienceId, ra, dec, width, height, units):
+        ''' Get the image specified by id centered on (ra, dec) with width and height dimensions.
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Line too long.

File: python/lsst/dax/imgserv/locateImage.py [hunk 5]
```diff
@@ -394,23 +381,22 @@ def getImageIdsFromScienceId(self, scienceId):
                 valid = False
             ids[key] = value
         return ids, valid
-
+
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: No trailing spaces.

File: python/lsst/dax/imgserv/locateImage.py [hunk 6]
```diff
@@ -422,6 +408,23 @@ def _getMetadata(self, butler, qResults):
             metadata = butler.get(self.getImageDatasetMd(), tract=tract, patch=patch, filter=filterName)
             return metadata
 
+
+    def _getImageCutoutFromScienceId(self, scienceId, ra, dec, width, height, units):
+        ''' Get the image specified by id centered on (ra, dec) with width and height dimensions.
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Line too long.

File: python/lsst/dax/imgserv/locateImage.py [hunk 7]
```diff
@@ -430,43 +433,19 @@ def _cutoutBoxPixels(srcImage, xyCenter, width, height, log):
...
-    # Cut the sub image out of the image. See -
-    # https://lsst-web.ncsa.illinois.edu/doxygen/x_masterDoxyDoc/afw_sec_py_image.html
+    src_box = srcImage.getBBox()
+    co_box = afwGeom.Box2I(xyCenter, afwGeom.Extent2I(int(width), int(height)))
+    co_box.clip(src_box)
+    pixULX = co_box.getBeginX()
+    pixEndX = co_box.getEndX()
+    pixULY = co_box.getBeginY()
+    pixEndY = co_box.getEndY()
+    log.debug("co_box pixULX={} pixEndX={} pixULY={} pixEndY={}".format(pixULX, pixEndX,
```
Comment (jgates108) [inline python/lsst/dax/imgserv/locateImage.py]: Line too long.

File: python/lsst/dax/imgserv/imageREST_v0.py [hunk 8]
```diff
@@ -226,6 +233,14 @@ def getIDeepCoaddCutout():
 def getIDeepCoaddCutoutPixel():
     return _getICutout(request, W13DeepCoaddDb, 'pixel')
 
+# this will handle something like:
+# GET /image/v0/deepCoadd/cutout?ra=19.36995&dec=-0.3146widthAng=30.0&heightAng=45.0
+# GET /image/v0/deepCoadd/cutout?ra=19.36995&dec=-0.3146xx&widthPix=100&heightPix=100
```
Comment (jgates108) [inline python/lsst/dax/imgserv/imageREST_v0.py]: Technically, these are too long, but I don't like the idea of splitting the example URL across 2 lines.

File: python/lsst/dax/imgserv/locateImage.py (hunk 1 above)
Comment (kennywlo) [inline python/lsst/dax/imgserv/locateImage.py]: Fixed.

File: python/lsst/dax/imgserv/locateImage.py (hunk 5 above)
Comment (kennywlo) [inline python/lsst/dax/imgserv/locateImage.py]: Fixed

File: python/lsst/dax/imgserv/locateImage.py (hunk 7 above)
Comment (kennywlo) [inline python/lsst/dax/imgserv/locateImage.py]: Fixed

**Pull request merged**